### ISSUE TITLE
Why the accuracy rate is so low from your model?

### ISSUE THREAD

**Issue by striversist** [da03/Attention-OCR#12]
Hi guys,

I've download the evaluation_data and model you supported, and verify the test result by the command 
`python src/launcher.py --phase=test --visualize --data-path=evaluation_data/svt/test.txt --data-base-dir=evaluation_data/svt --log-path=log.txt --load-model --model-dir=model --output-dir=results --target-embedding-size=10`  
(BTW. should add `--target-embedding-size=10` option or else it will fails by "Assign requires shapes of both tensors to match. lhs shape= [39,20] rhs shape= [39,10]")

But the result log.txt shows "188.780794 out of 647 correct" which means only 29% accuracy rate you get. Why is it so bad? Or Did I something wrong?

**Comment by gorinars** [comment]
I had the same issue whith exactly 188.7
It was solved  by dow-grading keras to v1.0.1. 

Maybe authors could add a note in readme, as something goes wrong in the decoder with new keras version

**Comment by striversist** [comment]
Hi gorinars, thanks!
After I downgraded Keras to v1.01, the result shows "437.192016 out of 647 correct" which means 68% accuracy. Much better!

But I have to say, it's still very poor comparing to other paper's result.
Hope the authors can tell us the best result they got.

**Comment by gorinars** [comment]
I agree. There was partial explanation from the authors here https://github.com/da03/Attention-OCR/issues/3

I tried several setups, but unfortunately was not able to reproduce state-of-the art 80% on SVT data set

**Comment by LinJM** [comment]
@gorinars  Now, what is your highest accuracy?

**Comment by LinJM** [comment]
@gorinars @striversist Do you know what is the point of the  downgrade the version of keras to 1.0.1?

**Comment by gorinars** [comment]
I came up with 68 percent on svt. No idea why downgrading helps. It just workerd

**Comment by striversist** [comment]
Hi gorinars, could you tell me what's your training parameters?
Strange, I can't get the accuracy by self training. 
See my new issue #14

**Comment by wanghaisheng** [comment]
@da03 
follow instructions under "Test and visualize attention results"
```
data-clear@dataclear:~/dev/tensorflow-playground/Attention-OCR$ docker run  -it --rm --name  attention-ocr  -v `pwd`:/home/work  dc/tensorflow-playground  /bin/bash

root@bdcd70cd18b0:/home/work# python -c "import keras; print keras.__version__"
Using TensorFlow backend.
1.1.1
root@bdcd70cd18b0:/home/work# python -c "import tensorflow as tf; print tf.__version__"
0.11.0
```

```
#!/usr/bin/env sh
python src/launcher.py --phase=test --visualize --data-path=evaluation_data/svt/test.txt --data-base-dir=evaluation_data/svt --log-path=log.txt --load-model --model-dir=model --output-dir=results 
```

what i got is really confusing
```
 perplexity: 13.937011
2016-11-25 05:29:14,096 root  INFO     21.000000 out of 634 correct
2016-11-25 05:29:14,277 root  INFO     step_time: 0.179516, loss: 2.263902, step perplexity: 9.620557
2016-11-25 05:29:14,286 root  INFO     21.000000 out of 635 correct
2016-11-25 05:29:14,417 root  INFO     step_time: 0.128888, loss: 3.065607, step perplexity: 21.447473
2016-11-25 05:29:14,423 root  INFO     21.000000 out of 636 correct
2016-11-25 05:29:14,544 root  INFO     step_time: 0.120134, loss: 3.372941, step perplexity: 29.164180
2016-11-25 05:29:14,548 root  INFO     21.000000 out of 637 correct
2016-11-25 05:29:14,666 root  INFO     step_time: 0.117341, loss: 2.897969, step perplexity: 18.137276
2016-11-25 05:29:14,669 root  INFO     21.000000 out of 638 correct
2016-11-25 05:29:14,794 root  INFO     step_time: 0.123457, loss: 2.392969, step perplexity: 10.945946
2016-11-25 05:29:14,797 root  INFO     21.000000 out of 639 correct
2016-11-25 05:29:14,896 root  INFO     step_time: 0.097963, loss: 2.654452, step perplexity: 14.217194
2016-11-25 05:29:14,897 root  INFO     21.000000 out of 640 correct
2016-11-25 05:29:15,021 root  INFO     step_time: 0.122504, loss: 3.273077, step perplexity: 26.392437
2016-11-25 05:29:15,025 root  INFO     21.000000 out of 641 correct
2016-11-25 05:29:15,180 root  INFO     step_time: 0.151728, loss: 1.636524, step perplexity: 5.137279
2016-11-25 05:29:15,188 root  INFO     21.000000 out of 642 correct
2016-11-25 05:29:15,367 root  INFO     step_time: 0.177205, loss: 1.791313, step perplexity: 5.997324
2016-11-25 05:29:15,372 root  INFO     21.000000 out of 643 correct
2016-11-25 05:29:15,556 root  INFO     step_time: 0.181739, loss: 2.187555, step perplexity: 8.913390
2016-11-25 05:29:15,564 root  INFO     21.000000 out of 644 correct
2016-11-25 05:29:15,727 root  INFO     step_time: 0.161122, loss: 2.385032, step perplexity: 10.859412
2016-11-25 05:29:15,734 root  INFO     21.000000 out of 645 correct
2016-11-25 05:29:15,830 root  INFO     step_time: 0.095012, loss: 3.402723, step perplexity: 30.045791
2016-11-25 05:29:15,832 root  INFO     21.000000 out of 646 correct
2016-11-25 05:29:15,966 root  INFO     step_time: 0.133317, loss: 3.468472, step perplexity: 32.087675
2016-11-25 05:29:15,973 root  INFO     21.000000 out of 647 correct

```

**Comment by da03** [comment]
Add flag --old-model-version. The reason is that keras changed its definition of parameters, and the model we provided is of an older version and needs addtional conversion.

**Comment by wanghaisheng** [comment]
```
root@bdcd70cd18b0:/home/work# ./test-visualize-attention-result-old.sh 
Using TensorFlow backend.
usage: launcher.py [-h] [--gpu-id GPU_ID] [--use-gru] [--phase {train,test}]
                   [--data-path DATA_PATH] [--data-base-dir DATA_BASE_DIR]
                   [--visualize] [--no-visualize] [--batch-size BATCH_SIZE]
                   [--initial-learning-rate INITIAL_LEARNING_RATE]
                   [--num-epoch NUM_EPOCH]
                   [--steps-per-checkpoint STEPS_PER_CHECKPOINT]
                   [--target-vocab-size TARGET_VOCAB_SIZE]
                   [--model-dir MODEL_DIR]
                   [--target-embedding-size TARGET_EMBEDDING_SIZE]
                   [--attn-num-hidden ATTN_NUM_HIDDEN]
                   [--attn-num-layers ATTN_NUM_LAYERS] [--load-model]
                   [--no-load-model] [--log-path LOG_PATH]
                   [--output-dir OUTPUT_DIR]
launcher.py: error: unrecognized arguments: --old-model-version
```

when i update source code through  git pull  ,it works now
```
2016-11-25 10:00:36,928 root  INFO     197.000000 out of 628 correct
2016-11-25 10:00:37,043 root  INFO     step_time: 0.112366, loss: 3.656601, step perplexity: 38.729476
2016-11-25 10:00:37,048 root  INFO     197.000000 out of 629 correct
2016-11-25 10:00:37,144 root  INFO     step_time: 0.093842, loss: 1.202595, step perplexity: 3.328745
2016-11-25 10:00:37,146 root  INFO     197.000000 out of 630 correct
2016-11-25 10:00:37,238 root  INFO     step_time: 0.091346, loss: 0.692998, step perplexity: 1.999701
2016-11-25 10:00:37,240 root  INFO     197.000000 out of 631 correct
2016-11-25 10:00:37,334 root  INFO     step_time: 0.092255, loss: 0.195785, step perplexity: 1.216265
2016-11-25 10:00:37,336 root  INFO     198.000000 out of 632 correct
2016-11-25 10:00:37,426 root  INFO     step_time: 0.088780, loss: 4.159286, step perplexity: 64.025793
2016-11-25 10:00:37,428 root  INFO     198.000000 out of 633 correct
2016-11-25 10:00:37,541 root  INFO     step_time: 0.111132, loss: 0.520379, step perplexity: 1.682666
2016-11-25 10:00:37,544 root  INFO     198.000000 out of 634 correct
2016-11-25 10:00:37,642 root  INFO     step_time: 0.094887, loss: 2.293780, step perplexity: 9.912332
2016-11-25 10:00:37,646 root  INFO     198.000000 out of 635 correct
2016-11-25 10:00:37,759 root  INFO     step_time: 0.111603, loss: 0.103224, step perplexity: 1.108739
2016-11-25 10:00:37,761 root  INFO     199.000000 out of 636 correct
2016-11-25 10:00:37,865 root  INFO     step_time: 0.102322, loss: 0.360526, step perplexity: 1.434083
2016-11-25 10:00:37,868 root  INFO     199.000000 out of 637 correct
2016-11-25 10:00:37,983 root  INFO     step_time: 0.111453, loss: 1.082935, step perplexity: 2.953334
2016-11-25 10:00:37,988 root  INFO     199.000000 out of 638 correct
2016-11-25 10:00:38,141 root  INFO     step_time: 0.151684, loss: 1.953241, step perplexity: 7.051508
2016-11-25 10:00:38,148 root  INFO     199.000000 out of 639 correct
2016-11-25 10:00:38,256 root  INFO     step_time: 0.106356, loss: 0.196389, step perplexity: 1.217000
2016-11-25 10:00:38,259 root  INFO     200.000000 out of 640 correct
2016-11-25 10:00:38,372 root  INFO     step_time: 0.112090, loss: 0.869567, step perplexity: 2.385877
2016-11-25 10:00:38,376 root  INFO     200.000000 out of 641 correct
2016-11-25 10:00:38,485 root  INFO     step_time: 0.107597, loss: 3.044632, step perplexity: 21.002300
2016-11-25 10:00:38,488 root  INFO     200.000000 out of 642 correct
2016-11-25 10:00:38,641 root  INFO     step_time: 0.151425, loss: 1.001689, step perplexity: 2.722876
2016-11-25 10:00:38,647 root  INFO     200.000000 out of 643 correct
2016-11-25 10:00:38,742 root  INFO     step_time: 0.094533, loss: 0.338087, step perplexity: 1.402263
2016-11-25 10:00:38,744 root  INFO     200.000000 out of 644 correct
2016-11-25 10:00:38,843 root  INFO     step_time: 0.093424, loss: 2.399895, step perplexity: 11.022021
2016-11-25 10:00:38,848 root  INFO     200.000000 out of 645 correct
2016-11-25 10:00:38,959 root  INFO     step_time: 0.108594, loss: 0.217461, step perplexity: 1.242917
2016-11-25 10:00:38,963 root  INFO     201.000000 out of 646 correct
2016-11-25 10:00:39,053 root  INFO     step_time: 0.088361, loss: 3.017737, step perplexity: 20.444971
2016-11-25 10:00:39,055 root  INFO     201.000000 out of 647 correct

```

**Comment by da03** [comment]
Yep I'm afraid that's indeed the exact match accuracy that the provided model can achieve. The model structure is not heavily tuned, and it's not trained until convergence. From my experience, using vanilla sgd with learning rate decay and gradient clipping can achieve a better optimum than adam (although we need to tune initial learning rate, decay and clipping threshold in that case). My repository torch-attention-ocr can train a model acieving state-of-the-art performance, though it's written in torch and does not come with a documentation at this time point.

**Comment by wanghaisheng** [comment]
thx for your help ,i will try torch  later

**Comment by lipond** [comment]
@striversist this problem solved now?

**Comment by ganliqiang** [comment]
hi @gorinars @striversist  @da03  i got the same result with yours,but when i check the results folds,i only find about 170 fold in correct fold which means accuracy is below 30%..,i have downgraded the keras version.when i use the train dataset to evaluate ,the output is 8700 out of 10000,but in correct fold ,only have 5000folds,that is very stange ,so do you have the same problem?thanks  a lot .
